### PR TITLE
JDK-8313796: AsyncGetCallTrace crash on unreadable interpreter method pointer

### PR DESCRIPTION
**Pull request by richardstartin**
Makes AsyncGetCallTrace more resilient to method unloading.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8313796](https://bugs.openjdk.org/browse/JDK-8313796): AsyncGetCallTrace crash on unreadable interpreter method pointer (**Bug** - P4)

### Reviewers
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to [86813482](https://git.openjdk.org/jdk/pull/15193/files/868134827c063f1160a2be97555f964eefb51034)
 * [Andrew Haley](https://openjdk.org/census#aph) (@theRealAph - **Reviewer**) ⚠️ Review applies to [6a257d4b](https://git.openjdk.org/jdk/pull/15193/files/6a257d4b6e5ea2ef3437a413c8c8635fb5e81b3f)
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**) ⚠️ Review applies to [86813482](https://git.openjdk.org/jdk/pull/15193/files/868134827c063f1160a2be97555f964eefb51034)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15193/head:pull/15193` \
`$ git checkout pull/15193`

Update a local copy of the PR: \
`$ git checkout pull/15193` \
`$ git pull https://git.openjdk.org/jdk.git pull/15193/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15193`

View PR using the GUI difftool: \
`$ git pr show -t 15193`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15193.diff">https://git.openjdk.org/jdk/pull/15193.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15193#issuecomment-1670057445)

### CODE REVIEW

**Review by coleenp — CHANGES_REQUESTED** [review]
Can you add a new function like frame::interpreter_frame_method_safe() or frame::safe_interpreter_frame_method() and put this code in shared/runtime ?

**Review by theRealAph — APPROVED** [review]
Great, thanks.

File: src/hotspot/share/runtime/frame.cpp [hunk 1]
```diff
@@ -301,6 +302,18 @@ bool frame::is_entry_frame_valid(JavaThread* thread) const {
   return (jfa->last_Java_sp() > sp());
 }
 
+Method* frame::safe_interpreter_frame_method() const {
+  Method** m_addr = interpreter_frame_method_addr();
+  if (m_addr == nullptr) {
+    return nullptr;
+  }
+  Method* m = (Method*) SafeFetchN((intptr_t*) m_addr, 0);
+  if (m == 0) {
```
Comment (tstuefe) [inline src/hotspot/share/runtime/frame.cpp:311]: Small nits:
- nullptr instead of 0 
- you can just return m. Its either nullptr or the Method.

**Review by tstuefe — APPROVED** [review]
Small nit, otherwise good.

File: src/hotspot/share/runtime/frame.cpp (hunk 1 above)
Comment (richardstartin) [reply src/hotspot/share/runtime/frame.cpp]: This didn't compile: `error: cannot convert 'std::nullptr_t' to 'intptr_t' {aka 'long int'}` - I put it back to the literal 0.

Comment (parttimenerd) [reply src/hotspot/share/runtime/frame.cpp]: `(void*)nullptr` is probably best, with `0` it is not directly obvious, that it is a pointer.

**Comment by richardstartin** [comment]
Thanks for the helpful reviews and for sponsoring

**Pull request closed without merging**